src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,540 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use tokio::sync::watch;
+use tokio::time::{sleep_until, Duration, Instant};
+
+const FIVE_MINUTES: Duration = Duration::from_secs(300);
+const TEN_SECONDS: Duration = Duration::from_secs(10);
+
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    rx_token: watch::Receiver<Option<Result<Token>>>,
+    _phantom: std::marker::PhantomData<T>,
+}
```

Comment:
I think you can simplify this to:

```suggestion
#[derive(Debug)]
pub(crate) struct TokenCache
{
    rx_token: watch::Receiver<Option<Result<Token>>>,
}
```

and then use the generic `T` parameter only in `new()`...